TombEngine/Objects/TR3/Entity/WaspMutant.cpp

```diff
@@ -0,0 +1,289 @@
+		particle.fadeToBlack = 2;
+		particle.sLife =
+		particle.life = 8;
+
+		particle.extras = 0;
+		particle.dynamic = -1;
+
+		particle.x = Random::GenerateInt(-8, 8);
+		particle.y = Random::GenerateInt(-8, 8);
+		particle.z = Random::GenerateInt(-64, 64);
+
+		particle.xVel = Random::GenerateInt(-32, 32);
+		particle.yVel = Random::GenerateInt(-32, 32);
+		particle.zVel = Random::GenerateInt(-32, 32);
+		particle.friction = 2 | (2 << 4);
+
+		particle.flags = SP_SCALE | SP_ITEM | SP_NODEATTACH | SP_DEF;
+		particle.gravity = particle.maxYvel = 0;
+		particle.scalar = 3;
+
+		particle.size =
+		particle.sSize = Random::GenerateFloat(PARTICLE_SIZE_MIN, PARTICLE_SIZE_MAX);
+		particle.dSize = particle.sSize / 2;
+	}
+
+	static void SpawnWaspMutantVenomSackEffects(ItemInfo& item, int itemNumber)
+	{
+		constexpr auto PARTICLE_EFFECT_COUNT = 2;
+
+		if (item.ItemFlags[0] < 0)
+			item.ItemFlags[0] = 0;
+
+		// Spawn light.
+		auto pos = GetJointPosition(&item, WaspVenomSackBite.meshNum, WaspVenomSackBite.Position);
+		TriggerDynamicLight(
+			pos.x, pos.y, pos.z, item.ItemFlags[0],
+			WaspVenomSackLightColor.x * UCHAR_MAX,
+			WaspVenomSackLightColor.y * UCHAR_MAX,
+			WaspVenomSackLightColor.z * UCHAR_MAX);
+
+		// Spawn particles.
+		for (int i = 0; i < PARTICLE_EFFECT_COUNT; i++)
+			SpawnWaspMutantVenomSackParticle(itemNumber);
+	}
+
+	void InitialiseWaspMutant(short itemNumber)
+	{
+		auto& item = g_Level.Items[itemNumber];
+
+		InitialiseCreature(itemNumber);
+		SetAnimation(&item, WASP_STATE_IDLE);
+		item.ItemFlags[0] = WASP_VENOM_SACK_LIGHT_POWER;
+	}
+
+	void WaspMutantControl(short itemNumber)
+	{
+		if (!CreatureActive(itemNumber))
+			return;
+
+		auto& item = g_Level.Items[itemNumber];
+		auto& creature = *GetCreatureInfo(&item);
+
+		short headingAngle = 0;
+
+		if (item.HitPoints <= 0)
+		{
+			switch (item.Animation.ActiveState)
+			{
+			case WASP_STATE_FALL:
+				if (item.Pose.Position.y >= item.Floor)
+				{
+					item.Animation.TargetState = WASP_STATE_DEATH;
+					item.Animation.IsAirborne = false;
+					item.Animation.Velocity.y = 0.0f;
+					item.Pose.Position.y = item.Floor;
+				}
+
+				break;
+
+			case WASP_STATE_DEATH:
+				if (item.ItemFlags[0] > 0 && item.ItemFlags[1] == 0)
+				{
+					item.ItemFlags[0]--;
+					item.ItemFlags[1] = 6;
+				}
+				else
+				{
+					item.ItemFlags[1]--;
+				}
+
+				item.Pose.Position.y = item.Floor;
+				break;
+
+			default:
+				SetAnimation(&item, WASP_STATE_FALL);
+				item.Animation.IsAirborne = true;
+				item.Animation.Velocity = Vector3::Zero;
+				break;
+			}
+
+			item.Pose.Orientation.x = 0;
+		}
+		else
+		{
+			if (item.AIBits)
+				GetAITarget(&creature);
+
+			AI_INFO ai;
+			CreatureAIInfo(&item, &ai);
+
+			GetCreatureMood(&item, &ai, true);
+			CreatureMood(&item, &ai, true);
+
+			headingAngle = CreatureTurn(&item, creature.MaxTurn);
+
+			switch (item.Animation.ActiveState)
+			{
+			case WASP_STATE_IDLE:
+				item.Pose.Position.y = item.Floor;
```

Comment:
Is this line needed? 

The fly to landing state sets the final pos.y and changes to this land idle state. Here the position is not changing so maybe this safety line is not needed?